### PR TITLE
Feature edit extern PROTO source

### PR DESCRIPTION
**Pull request by stefaniapedrazzi**
PROTO models defined as EXTERNPROTO can only be open in read-only mode (see #4727).
This PR added a new item in the Context Menu of PROTO nodes that:
* first checks if the current project path is valid and asks the user to relocate the project if needed
* copies the EXTERNPROTO file in the current projects (in case the file already exists a message dialog is shown)
* opens the local copy of the PROTO
* removes the EXTERNPROTO url from the WbProtoManager list
* prints a warning telling the user to save and reload the world

As for any change to PROTO models, the user has to reload the world to make them effective.
However, given that the EXTERNPROTO location is written in the world file, the user also has to save the file.

I renamed the context menu items to make them more consistent, here is the status in this PR:
* Local PROTO:
    * `Edit PROTO Source`
* Extern PROTO:
    *  `Edit PROTO Source`
    *  `View PROTO Source`

Remaining issues:
* [x] after reload (without quitting Webots) the EXTERNPROTO is still loaded instead of the local file even if the world doesn't contain any longer its definition ( #4773)
* [x] after restarting Webots, the local PROTO file cannot be loaded (see https://github.com/cyberbotics/webots/issues/4730#issuecomment-1168314217)

### CODE REVIEW

**Comment by ad-daniel** [comment]
Just FYI, I've left the importVrml functionality because it is still required to import wbo files (from the `addnode > import`) and exported using the `right-click proto > export` button. So we're removing all of it?

**Comment by stefaniapedrazzi** [comment]
> Just FYI, I've left the importVrml functionality because it is still required to import wbo files (from the addnode > import) and exported using the right-click proto > export button. So we're removing all of it?

I will check again.
But it didn't seem to be used anywhere.

**Comment by stefaniapedrazzi** [comment]
I confirm that the `WbMainWindow::importVrml` function is not used at all.
`WBO`(Webots objects) nodes are not  VRML files and can still be imported correctly.

**Comment by stefaniapedrazzi** [comment]
Note that you cannot see the PROTO modifications after reloading the world: will be fixed in https://github.com/cyberbotics/webots/issues/4773.

**Review by omichel — CHANGES_REQUESTED** [review]

File: src/webots/gui/WbMainWindow.cpp [hunk 1]
```diff
@@ -2290,12 +2258,54 @@ void WbMainWindow::updateProjectPath(const QString &oldPath, const QString &newP
...
+    const QString &protoFilePath = WbNetwork::instance()->get(fileName);
+    const QString protoFileName(QFileInfo(fileName).fileName());
+    if (modify && protoFileName.endsWith(".proto", Qt::CaseInsensitive)) {
+      if (WbMessageBox::question(tr("Your are trying to modify a remote PROTO file.") + "\n" +
+                                   tr("The PROTO file will be copied in the current project folder.") + "\n\n" +
+                                   tr("Do you want to continue it?"),
+                                 this, tr("Modify remote PROTO model")) == QMessageBox::Cancel)
+        return;
+
+      QString protosPath = WbProject::current()->path() + "protos";
```
[the remark this answers is not in the document]
Comment (stefaniapedrazzi) [reply src/webots/gui/WbMainWindow.cpp]: This variable cannot be const.

[the remark this answers is not in the document]
Comment (omichel) [reply src/webots/gui/WbMainWindow.cpp:2277]: Sorry, my bad.

**Review by omichel — APPROVED** [review]
Thank you.
It works nicely as expected.
However, I believe that when we download a PROTO to make it local, the world should change to refer to the local PROTO. So, the EXTERNPROTO list should be changed to point to the local PROTO instead of the remote one, and the world should be marked as modified (need to save) so that once we save and reload the world, the local PROTO is used instead of the remote one. I understood this is going to be addressed in #4773.

**Comment by stefaniapedrazzi** [comment]
After https://github.com/cyberbotics/webots/issues/4773, the new copied PROTO will be used by the world but only after saving and reloading the world.

Currently, after copying the PROTO, Webots internal PROTO URL is already changed to the local one.
To change the file and update the EXTERNPROTO declaration, we need to save the world. Currently this step is let to the user because but I can change it so that it is done automatically: just a search and replace in the world file.

However, as it is always the case modifications to the copied PROTO will only be applied after the world is reloaded.
I don't change this otherwise it will produce inconsistent behavior.

For the same reason, if we decide change automatically update EXTERNPROTO declaration, I would not set the world as modified given that this is not the expected behavior when a PROTO file is changed.

**Comment by omichel** [comment]
Thank you.
I tested it with the youbot.wbt example, and after editing the PROTO, saving and reloading the world, I got the following errors (which are not very clear to understand):
```
ERROR: Skipped PROTO 'Youbot' as it is not available at: C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/BodyMesh.proto.
ERROR: Skipped PROTO 'Youbot' as it is not available at: C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/Arm0Mesh.proto.
ERROR: Skipped PROTO 'Youbot' as it is not available at: C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/Arm1Mesh.proto.
ERROR: Skipped PROTO 'Youbot' as it is not available at: C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/Arm2Mesh.proto.
ERROR: Skipped PROTO 'Youbot' as it is not available at: C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/Arm3Mesh.proto.
ERROR: Skipped PROTO 'Youbot' as it is not available at: C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/Arm4Mesh.proto.
ERROR: Skipped PROTO 'Youbot' as it is not available at: C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/Arm5Mesh.proto.
ERROR: Skipped PROTO 'Youbot' as it is not available at: C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/FingerMesh.proto.
ERROR: Skipped PROTO 'Youbot' as it is not available at: C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/InteriorWheel.proto.
ERROR: Skipped PROTO 'Youbot' as it is not available at: C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/ExteriorWheel.proto.
ERROR: Skipped PROTO 'Youbot' as it is not available at: C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/PlateMesh.proto.
```

If I quit Webots and restart, I am getting even more warnings, even though the robot appears to be loaded in the 3D view minus the textures which are missing:
```
ERROR: Skipped PROTO 'Youbot' as it is not available at: C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/BodyMesh.proto.
ERROR: Skipped PROTO 'Youbot' as it is not available at: C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/Arm0Mesh.proto.
ERROR: Skipped PROTO 'Youbot' as it is not available at: C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/Arm1Mesh.proto.
ERROR: Skipped PROTO 'Youbot' as it is not available at: C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/Arm2Mesh.proto.
ERROR: Skipped PROTO 'Youbot' as it is not available at: C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/Arm3Mesh.proto.
ERROR: Skipped PROTO 'Youbot' as it is not available at: C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/Arm4Mesh.proto.
ERROR: Skipped PROTO 'Youbot' as it is not available at: C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/Arm5Mesh.proto.
ERROR: Skipped PROTO 'Youbot' as it is not available at: C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/FingerMesh.proto.
ERROR: Skipped PROTO 'Youbot' as it is not available at: C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/InteriorWheel.proto.
ERROR: Skipped PROTO 'Youbot' as it is not available at: C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/ExteriorWheel.proto.
ERROR: Skipped PROTO 'Youbot' as it is not available at: C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/PlateMesh.proto.
WARNING: Youbot (PROTO) > BodyMesh (PROTO) > Shape > KukaAlphaAppearance (PROTO) > ImageTexture: Texture file could not be read: 'C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/textures/kuka_alpha.png'
WARNING: Youbot (PROTO) > DEF ARM Solid > HingeJoint > Solid > Arm1Mesh (PROTO) > Shape > KukaBlackAppearance (PROTO) > ImageTexture: Texture file could not be read: 'C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/textures/kuka_black.png'
WARNING: Youbot (PROTO) > DEF ARM Solid > HingeJoint > Solid > HingeJoint > Solid > HingeJoint > Solid > Arm3Mesh (PROTO) > Shape > KukaBlackAppearance (PROTO) > ImageTexture: Texture file could not be read: 'C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/textures/kuka_black.png'
WARNING: Youbot (PROTO) > DEF ARM Solid > HingeJoint > Solid > HingeJoint > Solid > HingeJoint > Solid > HingeJoint > Solid > Arm4Mesh (PROTO) > Shape > KukaBlackAppearance (PROTO) > ImageTexture: Texture file could not be read: 'C:/Users/Olivi/AppData/Local/Programs/Webots/projects/robots/kuka/youbot/protos/textures/kuka_black.png'
INFO: youbot: Starting controller: C:\Users\Olivi\Desktop\youbot\controllers\youbot\youbot.exe demo
```

**Comment by omichel** [comment]
I had to manually fix the PROTO file by searching and replacing `webots://` by `https://raw.githubusercontent.com/cyberbotics/webots/9d9d2686d2242f310a0d03c3e08a0588a9576b6e/` in the EXTERNPROTO declarations (and we should do it for any `webots://` URL.

**Comment by stefaniapedrazzi** [comment]
> I believe this should be done automatically after we copy the PROTO file into the local protos folder.

Not sure.
If they are working in the original file, while they should not work (given that are correct URLs) when the file is copied.
I would first check why they are not found after the copy.

**Comment by ad-daniel** [comment]
I confirm that when a PROTO is created in the local project (from a remote one for example), the `webots://` urls need to be converted to target the current version `raw.github../R2022b/`. Something analogous is done for the [export PROTO](https://github.com/cyberbotics/webots/blob/bbbbd2bfb4593096dfbf0f19207eee59115ae5b0/src/webots/vrml/WbProtoManager.cpp#L509-L560) functionality (when exporting from the AddNode window), but now that you mention it, I believe that functionality only converts the EXTERNPROTO references whereas it should apply to ALL `webots://` urls, textures included, so there is likely a bug in that as well

**Comment by ad-daniel** [comment]
I'll address both issues

**Comment by omichel** [comment]
I am still getting this kind of error when editing the YouBot proto file:
```
ERROR: Error downloading EXTERNPROTO 'BodyMesh': Cannot download 'https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/kuka/youbot/protos/BodyMesh.proto', error code: 203: Error transferring https://raw.githubusercontent.com/cyberbotics/webots/R2022b/projects/robots/kuka/youbot/protos/BodyMesh.proto - server replied: Not Found
```
However, I believe it will be resolved when we publish the tag for R2022b.

File: src/webots/gui/WbMainWindow.cpp [hunk 2]
```diff
@@ -2290,12 +2258,75 @@ void WbMainWindow::updateProjectPath(const QString &oldPath, const QString &newP
...
+    if (modify && protoFileName.endsWith(".proto", Qt::CaseInsensitive)) {
+      if (WbMessageBox::question(
+            tr("You are trying to modify a remote PROTO file.") + "\n" +
+              tr("The PROTO file will be copied in the current project folder.") + "\n" +
+              tr("You should save and reload the world file, so that it refers to this local PROTO file.") + "\n\n" +
+              tr("Do you want to proceed?"),
+            this, tr("Modify remote PROTO model")) == QMessageBox::Cancel)
+        return;
+
+      QString protosPath = WbProject::current()->path() + "protos";
```
Comment (omichel) [inline src/webots/gui/WbMainWindow.cpp:2279]:
```suggestion
      const QString protosPath = WbProject::current()->protosPath();
```

**Review by omichel — CHANGES_REQUESTED** [review]
Just 2 minor things.

**Review by omichel — APPROVED** [review]
Thank you.

**Pull request merged**